state-chain/pallets/cf-cfe-event-emitter/src/lib.rs

```diff
@@ -0,0 +1,282 @@
+	pub from_epoch: EpochIndex,
+	pub to_epoch: EpochIndex,
+	pub key_to_share: <BitcoinCrypto as ChainCrypto>::AggKey,
+	pub sharing_participants: BTreeSet<T::ValidatorId>,
+	pub receiving_participants: BTreeSet<T::ValidatorId>,
+	pub new_key: <BitcoinCrypto as ChainCrypto>::AggKey,
+}
+
+#[derive(Clone, RuntimeDebug, PartialEq, Eq, Encode, Decode, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct TxBroadcastRequest<T: Chainflip, C: Chain> {
+	pub broadcast_id: BroadcastId,
+	pub nominee: T::ValidatorId,
+	pub payload: C::Transaction,
+}
+
+#[derive(Clone, RuntimeDebug, Encode, Decode, PartialEq, Eq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub enum CfeEvent<T: Chainflip> {
+	EthThresholdSignatureRequest(ThresholdSignatureRequest<T, EvmCrypto>),
+	DotThresholdSignatureRequest(ThresholdSignatureRequest<T, PolkadotCrypto>),
+	BtcThresholdSignatureRequest(ThresholdSignatureRequest<T, BitcoinCrypto>),
+	EthKeygenRequest(KeygenRequest<T>),
+	DotKeygenRequest(KeygenRequest<T>),
+	BtcKeygenRequest(KeygenRequest<T>),
+	BtcKeyHandoverRequest(KeyHandoverRequest<T>),
+	EthTxBroadcastRequest(TxBroadcastRequest<T, Ethereum>),
+	DotTxBroadcastRequest(TxBroadcastRequest<T, Polkadot>),
+	BtcTxBroadcastRequest(TxBroadcastRequest<T, Bitcoin>),
+	PeerIdRegistered {
+		account_id: T::ValidatorId,
+		pubkey: Ed25519PublicKey,
+		port: Port,
+		ip: Ipv6Addr,
+	},
+	PeerIdDeregistered {
+		account_id: T::ValidatorId,
+		pubkey: Ed25519PublicKey,
+	},
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+
+	use cf_traits::Chainflip;
+
+	use super::*;
+
+	#[pallet::pallet]
+	#[pallet::storage_version(PALLET_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	#[pallet::disable_frame_system_supertrait_check]
+	pub trait Config: Chainflip {}
+
+	#[pallet::storage]
+	#[pallet::unbounded]
+	pub type CfeEvents<T: Config> =
+		StorageMap<_, Twox64Concat, BlockNumberFor<T>, Vec<CfeEvent<T>>>;
+}
+
+fn add_event<T: Config>(event: CfeEvent<T>) {
+	let number = frame_system::Pallet::<T>::block_number();
+
+	CfeEvents::<T>::mutate(number, |events| {
+		let events = events.get_or_insert(Vec::new());
+		events.push(event);
+	})
+}
+
+impl<T: Config> CfeEventEmitterForCrypto<T, EvmCrypto> for Pallet<T> {
+	fn keygen_request(
+		ceremony_id: CeremonyId,
+		epoch_index: EpochIndex,
+		participants: BTreeSet<T::ValidatorId>,
+	) {
+		add_event::<T>(CfeEvent::EthKeygenRequest(KeygenRequest {
+			ceremony_id,
+			epoch_index,
+			participants,
+		}));
+	}
+
+	// TODO: consider taking directly taking ThresholdSignatureRequest as parameter (though this
+	// would require cf-traits importing ThresholdSignatureRequest, which is a non-primitive type)
+	fn signature_request(
+		ceremony_id: CeremonyId,
+		epoch_index: EpochIndex,
+		key: <EvmCrypto as ChainCrypto>::AggKey,
+		signatories: BTreeSet<T::ValidatorId>,
+		payload: <EvmCrypto as ChainCrypto>::Payload,
+	) {
+		add_event::<T>(CfeEvent::EthThresholdSignatureRequest(ThresholdSignatureRequest {
```

Comment:
It ought to be possible to remove a lot of the duplication in the SCO relating to these requests, and I think this might make it a little more difficult? I think that's worth keeping in mind anyway - that when we do this and put these type instances in their own enum variant - we may lose the benefit of the generic to some extent. But this is the kind of thing I think just requires some poking around to see it all fit together - might not be so simple given we probably do want to have a single unified enum of `CfeEvent` to encompass all CFE events. as we already do here.

Perhaps duplication can be removed later with a macro if we deem it necessary / nice.